File: test-suite-python/src/test/python/micronaut/docs/propagation/MdcServiceSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property, Requires
from micronaut.http import MediaType
from micronaut.http.annotation import Controller, Get
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .MdcService import MdcService


@Requires(property="mdc.example.service.enabled")
@Controller("/mdc")
class MDCController:

    def __init__(self, mdcService: MdcService):
        self.mdcService = mdcService

    @Get(value="/test", produces=MediaType.TEXT_PLAIN)
    def test(self) -> str:
        return self.mdcService.createUser("Denis")


@Property(name="mdc.example.service.enabled", value=True)
@MicronautTest
class MdcServiceSpec:
    client: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def testFilterSpec(self) -> None:
        response = self.client.toBlocking().retrieve("/mdc/test")
        assert response.startswith("New user id: ")
        assert response.endswith(" name: Denis")

File: test-suite-scala/src/test/scala/io/micronaut/docs/propagation/MdcServiceSpec.scala
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.propagation

import io.micronaut.context.ApplicationContext
import io.micronaut.http.HttpRequest
import io.micronaut.http.client.HttpClient
import io.micronaut.runtime.server.EmbeddedServer
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

import scala.jdk.CollectionConverters.*

class MdcServiceSpec:

  @Test
  def testFilterSpec(): Unit =
    val embeddedServer = ApplicationContext.run(
      classOf[EmbeddedServer],
      Map[String, Object]("mdc.example.service.enabled" -> java.lang.Boolean.TRUE).asJava
    )
    try
      val client = HttpClient.create(embeddedServer.getURL)
      try
        val request = HttpRequest.GET[Any]("/mdc/test")
        val response = client.toBlocking.retrieve(request)
        assertTrue(response.startsWith("New user id: "))
        assertTrue(response.endsWith(" name: Denis"))
      finally
        client.close()
    finally
      embeddedServer.close()
